os/drivers/audio/alc1019.c

```diff
@@ -0,0 +1,1083 @@
+}
+
+/****************************************************************************
+ * Name: alc1019_set_i2s_samplerate
+ *
+ * Description:
+ *
+ ****************************************************************************/
+static void alc1019_set_i2s_samplerate(FAR struct alc1019_dev_s *priv)
+{
+        /* if no audio device object return */
+        if (!priv) {
+                auddbg("Error, Device's private data Not available\n");
+                return;
+        }
+
+        if (priv->inout) {
+                I2S_RXSAMPLERATE(priv->i2s, priv->samprate);
+        } else {
+                I2S_TXSAMPLERATE(priv->i2s, priv->samprate);
+        }
+}
+
+/****************************************************************************
+ * Name: alc1019_getcaps
+ *
+ * Description:
+ *   Get the audio device capabilities
+ *
+ ****************************************************************************/
+static int alc1019_getcaps(FAR struct audio_lowerhalf_s *dev, int type, FAR struct audio_caps_s *caps)
+{
+#if !defined(CONFIG_AUDIO_EXCLUDE_VOLUME) || !defined(CONFIG_AUDIO_EXCLUDE_TONE)
+	FAR struct alc1019_dev_s *priv = (FAR struct alc1019_dev_s *)dev;
+#endif
+	/* Validate the structure */
+
+	DEBUGASSERT(caps && caps->ac_len >= sizeof(struct audio_caps_s));
+	audvdbg("type=%d ac_type=%d\n", type, caps->ac_type);
+
+	/* Fill in the caller's structure based on requested info */
+
+	caps->ac_controls.w = 0;
+
+	switch (caps->ac_type) {
+	/* Caller is querying for the types of units we support */
+
+	case AUDIO_TYPE_QUERY:
+
+		/* Provide our overall capabilities.  The interfacing software
+		 * must then call us back for specific info for each capability.
+		 */
+
+		caps->ac_channels = ALC1019_CHANNELS;	/* Stereo output */
+
+		switch (caps->ac_subtype) {
+		case AUDIO_TYPE_QUERY:
+			/* We don't decode any formats!  Only something above us in
+			 * the audio stream can perform decoding on our behalf.
+			 */
+
+			/* The types of audio units we implement */
+			caps->ac_controls.b[0] = AUDIO_TYPE_OUTPUT | AUDIO_TYPE_FEATURE;
+
+			break;
+
+		default:
+			caps->ac_controls.b[0] = AUDIO_SUBFMT_END;
+			break;
+		}
+
+		break;
+
+	/* Provide capabilities of our INPUT & OUTPUT unit */
+	case AUDIO_TYPE_OUTPUT:
+		caps->ac_channels = ALC1019_CHANNELS;
+		switch (caps->ac_subtype) {
+		case AUDIO_TYPE_QUERY:
+			/* Report the Sample rates we support */
+			caps->ac_controls.b[0] = AUDIO_SAMP_RATE_TYPE_48K;
+			break;
+
+		case AUDIO_FMT_MP3:
+		case AUDIO_FMT_WMA:
+		case AUDIO_FMT_PCM:
+			break;
+
+		default:
+			break;
+		}
+
+		break;
+
+	/* Provide capabilities of our FEATURE units */
+
+	case AUDIO_TYPE_FEATURE:
+
+		/* If the sub-type is UNDEF, then report the Feature Units we support */
+
+		if (caps->ac_subtype == AUDIO_FU_UNDEF) {
+			/* Fill in the ac_controls section with the Feature Units we have */
+
+			caps->ac_controls.b[0] = AUDIO_FU_VOLUME | AUDIO_FU_BASS | AUDIO_FU_TREBLE;
+			caps->ac_controls.b[1] = AUDIO_FU_BALANCE >> 8;
```

Comment:
only AUDIO_FU_VOLUME is available.
so alc1019_configure also need to be updated